language/evm/examples/sources/ERC721.move

```diff
@@ -0,0 +1,212 @@
+
+    #[event]
+    struct Transfer {
+        from: address,
+        to: address,
+        tokenId: u128,
+    }
+
+    #[event]
+    struct Approval {
+        owner: address,
+        approved: address,
+        tokenId: u128,
+    }
+
+    #[event]
+    struct ApprovalForAll {
+        owner: address,
+        operator: address,
+        approved: bool,
+    }
+
+    #[storage]
+    /// Represents the state of this contract. This is located at `borrow_global<State>(self())`.
+    struct State has key {
+        owners: Table<u128, address>,
+        balances: Table<address, u128>,
+        tokenApprovals: Table<u128, address>,
+        operatorApprovals: Table<address, Table<address, bool>>,
+    }
+
+    #[create]
+    /// Constructor of this contract.
+    public fun create() {
+        // Initial state of contract
+        move_to<State>(
+            sign(self()),
+            State {
+                owners: Table::empty<u128, address>(),
+                balances: Table::empty<address, u128>(),
+                tokenApprovals: Table::empty<u128, address>(),
+                operatorApprovals: Table::empty<address, Table<address, bool>>(),
+            }
+        );
+    }
+
+    #[callable]
+    /// Count all NFTs assigned to an owner.
+    public fun balanceOf(owner: address): u128 acquires State {
+        let s = borrow_global_mut<State>(self());
+        *mut_balanceOf(s, owner)
+    }
+
+    #[callable]
+    /// Find the owner of an NFT.
+    public fun ownerOf(tokenId: u128): address acquires State {
+        let s = borrow_global_mut<State>(self());
+        *mut_ownerOf(s, tokenId)
+    }
+
+    #[callable(name=safeTransferFrom)] // Overloading `safeTransferFrom`
+    /// Transfers the ownership of an NFT from one address to another address.
+    public fun safeTransferFrom_with_data(from: address, to: address, tokenId: u128, data: vector<u8>) acquires State {
+        transferFrom(from, to, tokenId);
+        doSafeTransferAcceptanceCheck(from, to, tokenId, data);
+
+    }
+
+    #[callable]
+    /// Transfers the ownership of an NFT from one address to another address.
+    public fun safeTransferFrom(from: address, to: address, tokenId: u128) acquires State {
+        safeTransferFrom_with_data(from, to, tokenId, b"");
+    }
+
+    #[callable]
+    /// Transfer ownership of an NFT. THE CALLER IS RESPONSIBLE
+    ///  TO CONFIRM THAT `_to` IS CAPABLE OF RECEIVING NFTS OR ELSE
+    ///  THEY MAY BE PERMANENTLY LOST
+    public fun transferFrom(from: address, to: address, tokenId: u128) acquires State {
+        assert!(isApprovedOrOwner(sender(), tokenId), Errors::invalid_argument(0));
+        assert!(ownerOf(tokenId) == from, Errors::invalid_argument(0));
+        assert!(to != @0x0, Errors::invalid_argument(0));
+
+        // Clear approvals from the previous owner
+        approve(@0x0, tokenId);
+
+        let s = borrow_global_mut<State>(self());
+
+        let mut_balance_from = mut_balanceOf(s, from);
```

Comment:
`balances` currently has type `Table<address, u128>`. We could do `Table<address, Balance>`. If `Balance` is defined in the same module just to wrap a `u128` value, it may be unnecessary to do that. However, `Balance` is defined to guarantee some linearity for the value in it, it's worthing considering it! Let's think more about this point when we iterate these examples.